File: backend/services/search_service.py
import math
from typing import Dict, Any, List, Optional
from backend.database import STANDARDS_DB

class SearchService:
    @staticmethod
    def search(
        query: str,
        category: Optional[str] = None,
        mandatory_only: bool = False,
        page: int = 1,
        limit: int = 20
    ) -> Dict[str, Any]:
        clean_query = query.lower().strip()
        tokens = [t for t in clean_query.split() if len(t) > 1]

        scored_items = []

        for std in STANDARDS_DB:
            if category and category != "All":
                if category.lower() not in std.get("category", "").lower():
                    continue

            if mandatory_only and not (std.get("isMandatory") or std.get("mandatory")):
                continue

            score = 0
            matched = []

            code_lower = std.get("code", "").lower()
            title_lower = std.get("title", "").lower()
            summary_lower = std.get("summary", "").lower()

            # 1. Exact code match
            if code_lower == clean_query or clean_query in code_lower:
                score += 100
                matched.append("code_exact")

            # 2. Phrase in title
            if clean_query in title_lower:
                score += 50
                matched.append("title_phrase")

            # 3. Token matching
            for token in tokens:
                if token in code_lower:
                    score += 25
                    matched.append(f"code:{token}")
                if token in title_lower:
                    score += 15
                    matched.append(f"title:{token}")
                if token in summary_lower:
                    score += 8
                    matched.append(f"summary:{token}")
                if any(token in k.lower() for k in std.get("keywords", [])):
                    score += 10
                    matched.append(f"keyword:{token}")
                if any(token in b.lower() for b in std.get("businessTypes", [])):
                    score += 12
                    matched.append(f"business:{token}")

            # 4. Clause level match
            for clause in std.get("clauses", []):
                if clean_query in clause.get("content", "").lower():
                    score += 20
                    matched.append(f"clause:{clause.get('number')}")
                    break

            # 5. Mandatory boost
            if std.get("isMandatory") or std.get("mandatory"):
                score += 5

            if score > 0:
                scored_items.append((score, list(set(matched)), std))

        # Sort descending by score
        scored_items.sort(key=lambda x: x[0], reverse=True)

        total = len(scored_items)
        offset = (page - 1) * limit
        paginated = scored_items[offset : offset + limit]

        results = []
        for score, matched_kw, std in paginated:
            results.append({
                "id": std["id"],
                "code": std["code"],
                "title": std["title"],
                "year": std["year"],
                "category": std["category"],
                "division": std.get("division") or std.get("department", ""),
                "isMandatory": std.get("isMandatory", False),
                "scheme": std.get("scheme", "Scheme I (ISI Mark)"),
                "summary": std.get("summary", ""),
                "relevanceScore": int(score),
                "matchedKeywords": matched_kw
            })

        return {
            "results": results,
            "pagination": {
                "page": page,
                "limit": limit,
                "total": total,
                "totalPages": math.ceil(total / limit) if limit > 0 else 1
            }
        }

File: src/server/services/search.service.ts
import { STANDARDS_DATABASE, Standard } from "@/lib/standards-data";

export interface SearchOptions {
  query: string;
  category?: string;
  mandatoryOnly?: boolean;
  page?: number;
  limit?: number;
}

export interface SearchResultItem {
  id: string;
  code: string;
  title: string;
  year: number;
  category: string;
  division: string;
  isMandatory: boolean;
  scheme: string;
  summary: string;
  relevanceScore: number;
  matchedKeywords: string[];
}

export class SearchService {
  static search(opts: SearchOptions): {
    results: SearchResultItem[];
    pagination: { page: number; limit: number; total: number; totalPages: number };
  } {
    const { query, category, mandatoryOnly, page = 1, limit = 20 } = opts;
    const cleanQuery = query.toLowerCase().trim();
    const queryTokens = cleanQuery.split(/\s+/).filter((t) => t.length > 1);

    const scoredItems: Array<{ standard: Standard; score: number; matched: string[] }> = [];

    for (const std of STANDARDS_DATABASE) {
      // 1. Hard filters
      if (category && category !== "All") {
        if (!std.category.toLowerCase().includes(category.toLowerCase())) continue;
      }

      if (mandatoryOnly && !(std.isMandatory || std.mandatory)) {
        continue;
      }

      // 2. Multi-factor Scoring
      let score = 0;
      const matched: string[] = [];

      const codeLower = std.code.toLowerCase();
      const titleLower = std.title.toLowerCase();
      const summaryLower = std.summary.toLowerCase();

      // Exact code match (highest weight)
      if (codeLower === cleanQuery || codeLower.includes(cleanQuery)) {
        score += 100;
        matched.push("code_exact");
      }

      // Exact title match
      if (titleLower.includes(cleanQuery)) {
        score += 50;
        matched.push("title_phrase");
      }

      // Token matching across title, summary, keywords, and business types
      for (const token of queryTokens) {
        if (codeLower.includes(token)) {
          score += 25;
          matched.push(`code:${token}`);
        }
        if (titleLower.includes(token)) {
          score += 15;
          matched.push(`title:${token}`);
        }
        if (summaryLower.includes(token)) {
          score += 8;
          matched.push(`summary:${token}`);
        }
        if (std.keywords.some((k) => k.toLowerCase().includes(token))) {
          score += 10;
          matched.push(`keyword:${token}`);
        }
        if (std.businessTypes.some((b) => b.toLowerCase().includes(token))) {
          score += 12;
          matched.push(`business:${token}`);
        }
      }

      // Clause contents matching
      for (const clause of std.clauses) {
        if (clause.content.toLowerCase().includes(cleanQuery)) {
          score += 20;
          matched.push(`clause:${clause.number}`);
          break;
        }
      }

      // Mandatory boost for industrial searchers
      if (std.isMandatory || std.mandatory) {
        score += 5;
      }

      if (score > 0) {
        scoredItems.push({ standard: std, score, matched: Array.from(new Set(matched)) });
      }
    }

    // Sort by relevance score descending
    scoredItems.sort((a, b) => b.score - a.score);

    const total = scoredItems.length;
    const offset = (page - 1) * limit;
    const paginated = scoredItems.slice(offset, offset + limit);

    const results: SearchResultItem[] = paginated.map(({ standard: s, score, matched }) => ({
      id: s.id,
      code: s.code,
      title: s.title,
      year: s.year,
      category: s.category,
      division: s.division || s.department,
      isMandatory: s.isMandatory || s.mandatory,
      scheme: s.scheme,
      summary: s.summary,
      relevanceScore: Math.round(score),
      matchedKeywords: matched,
    }));

    return {
      results,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }
}
